src/Generator/Generators/CSharp/CSharpTextTemplate.cs

```diff
@@ -196,6 +198,80 @@ public void GenerateHeader()
             PopBlock();
         }
 
+
+        private class DelegateSignature
+        {
+            public string Signature { get; set; }
+            public bool SuppressWarning { get; set; }
+            public string CallingConvention { get; set; }
+            public string StrReturnType { get; set; }
+        }
+
+        private class OptionalOut<Type>
+        {
+            public List<Type> Result { get; set; }
+        }
+        private DelegateSignature GenerateDelegateSignature(Function function, bool suppressWarning, out CSharpTypePrinterResult retType, out string delegateName)
+        {
+            OptionalOut<string> allTypes = new OptionalOut<string>();
+            allTypes.Result = new List<string>();
+            var @params = GatherInternalParams(function, out retType, false, allTypes);
+
+            DelegateSignature ds = new DelegateSignature();
+            ds.CallingConvention = function.CallingConvention.ToInteropCallConv().ToString();
+            ds.SuppressWarning = suppressWarning;
+
+            allTypes.Result.Add(ds.CallingConvention);
+            allTypes.Result.Add(suppressWarning.ToString());
+            delegateName = string.Join("_", allTypes.Result);
+
+            ds.Signature =  string.Join(", ", @params);
+            ds.StrReturnType = retType.ToString();
+            return ds;
+        }
+
+        private Dictionary<string, DelegateSignature> AllDelegates = new Dictionary<string, DelegateSignature>();
+
+        private string GenerateDelegate(Method func, bool suppressWarning = true)
+        {
+            CSharpTypePrinterResult retType;
+            string strName;
+            DelegateSignature signature = GenerateDelegateSignature(func, suppressWarning, out retType, out strName);
+            if (func.ReturnType.Type.IsPrimitiveType(PrimitiveType.Void))
+                strName = "Func_" + strName;
+            else
+                strName = "Action_" + strName;
+            DelegateSignature delegateSign;
+            if (AllDelegates.TryGetValue(strName, out delegateSign))
+                return strName;
+
+            AllDelegates.Add(strName, signature);
+            return strName;
+        }
+        private string GenerateDelegate(Function func, bool suppressWarning = true)
+        {
+            return GenerateDelegate(func as Method, suppressWarning);
+        }
+
+        private void GenerateDelegatesClass()
+        {
+            if (AllDelegates.Count == 0)
+                return;
+            DelegateSignature ds;
```

Comment:
Please move these within the loop using "var":
var delegateName = e.Key;
var ds = e.Value;